examples/common/convert_c4.py

```diff
@@ -4,105 +4,359 @@
+                encoded = self.tokenizer(batch['text'], truncation=False, padding=False)
+                for iids in encoded['input_ids']:
+                    buffer['tokens'] = buffer.get('tokens', []) + iids
+                    while len(buffer['tokens']) >= self.max_length:
+                        concat_sample = {}
+                        concat_sample['tokens'] = buffer['tokens'][:self.
+                                                                max_length]
+                        buffer['tokens'] = buffer['tokens'][self.max_length:]
+
+                        yield {
+                            # convert to bytes to store in MDS binary format
+                            'tokens': np.asarray(concat_sample['tokens']).tobytes()
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+def build_hf_c4_dataset(
+        split: str,
+        mode: ConcatMode,
+        max_length: Optional[int] = None,
+        sep_text: Optional[str] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        expected_num_samples: Optional[int] = None) -> IterableDataset:
     """Collect the samples for this dataset split.
 
     Args:
         split (str): Split name.
+        mode (ConcatMode): NO_CONCAT, CONCAT_TEXT, or CONCAT_TOKENS
+        sep_text (str): if mode is CONCAT_TEXT, what string to use to separate samples
+        tokenizer (PreTrainedTokenizerBase): if mode is CONCAT_TOKENS, the tokenizer to use
+        expected_num_samples (int): _not yet implemented_ once we have generated this many samples, quit
 
     Returns:
         An IterableDataset.
     """
+    if mode == ConcatMode.NO_CONCAT:
+        dataset = ShardedC4(split, expected_num_samples=expected_num_samples)
+    elif mode == ConcatMode.CONCAT_TEXT:
+        dataset = ConcatC4(split=split,
+                           max_length=max_length,
+                           sep_text=sep_text,
+                           expected_num_samples=expected_num_samples)
+    else:
+        dataset = TokenizedC4(split=split,
+                              tokenizer=tokenizer,
+                              max_length=max_length,
+                              expected_num_samples=expected_num_samples)
+    return dataset
 
-    class ShardedC4(IterableDataset):
 
-        def __init__(self):
-            self.dataset = hf_datasets.load_dataset(path='c4',
-                                                    name='en',
-                                                    split=split,
-                                                    streaming=True)
+def _concat_sample_count(sample_count: int,
+                         mode: ConcatMode,
+                         max_length: Optional[int] = None) -> int:
+    emp_chars_per_sample = 2163  # measured for val split
+    est_tokens_per_samples = 540  # using est_char_per_token = 4
+    match mode:
```

Comment:
this forces us to python 3.10, so maybe lets just use if/else?

Reply:
good point on the version. it was finally a use case for matching with guards and I got excited